pkg/hsskey/hsskey.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package hsskey provides functionality for generating a key for unlocking
+// drives based on the following procedure:
+//  1. Via BMC, read a 32-byte secret seed known as the Host Secret Seed (HSS)
+//     using the OpenBMC IPMI blob transfer protocol
+//  2. Compute a password as follows:
+//     We get the deterministically computed 32-byte HDKF-SHA256 using:
+//     - salt: "SKM PROD_V2 ACCESS"
+//     - hss: 32-byte HSS
+//     - device identity: strings formed by concatenating the assembly serial
+//     number, the _ character, and the assembly part number.
+package hsskey
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"log"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/ipmi"
+	"github.com/u-root/u-root/pkg/ipmi/blobs"
+	"golang.org/x/crypto/hkdf"
+)
+
+const (
+	hostSecretSeedLen = 32
+
+	passwordSalt = "SKM PROD_V2 ACCESS"
```

Comment:
It seems dangerous to me that we're hard coding a password salt here in the `hsskey` package? Should we instead rely on the caller of `GenPassword` to  pass in the desired salt (or generate a random one, if that makes sense)?

Reply:
These tools are all for working with drives that have already been encrypted, not for setting up new ones, so generating a salt isn't really relevant here.  Passing in the salt from the caller would be reasonable from a design standpoint, but in practice that would just result in it being hardcoded to this same value on every caller's side, so I don't think it's preferable at the moment.

Reply:
> Passing in the salt from the caller would be reasonable from a design standpoint, but in practice that would just result in it being hardcoded to this same value on every caller's side, so I don't think it's preferable at the moment.

I think that passing it from the binary is reasonable, even if it's still hardcoded there, because the package `pkg/hsskey/hsskey.go` may be reused by another binary/package at a different time in the future and we want to be sure that other future users of this package don't wind up using the same salt by mistake.

Also, instead of hard coding it, would it be possible to change the salt to be a flag in `disk_unlock` binary (with the existing default value) instead?  Then whatever humans/robots are calling `disk_unlock` can be changed to set that flag to the current value, and once all the callers of the binary are migrated to use the flag we can remove the default value?

Reply:
Maybe it would be better to just rename `GenPassword` to something more specific, like `GenProdV2Password`.  It's not intended as a generic password creation function, but rather an implementation of a specific password-generation procedure that includes several arbitrary decisions beyond just the salt: the hash function, the key derivation function, the number of bytes used as the key...

We do have future aspirations to make this code more generically usable by separating out platform-specific (and potentially security sensitive) parts into some kind of configuration file, and utilize the OSS cryptographic algorithm components here based on that configuration.  Hopefully that will improve on the concerns here down the line.

Reply:
I agree that the phrase should be a flag. We can still set "our" phrase as the default value. But u-root is composed of tools that are theoretically and hopefully practically reusable by companies other than Google, and we should make that easier. So let's make a flag?

Reply:
I'm pushing back on this because I do not believe it's a parameter that will ever be changed from this value in practice.  Another organization trying to use the function would most likely either need to add other new parameter(s) anyways (to match their particular key generation procedure), or would just write their own version of the function.

If it's going to block this PR, I'll make the change, but to me it feels very YAGNI.

Reply:
> Another organization trying to use the function would most likely either need to add other new parameter(s) anyways (to match their particular key generation procedure), or would just write their own version of the function.

Does it make sense to abstract the organization-specific logic as a `type` that is passed to the package (which contains generic organization-independent code?)

`type KeyGenerator func() (error, key)` or something? 

I'm sorry but I lack context on what parts are specific to the organization, what are generic, etc.

If this is code that will never be used by any other organization, does it make sense to include it here in the public `u-root` package?   If it is something that can be useful by another organization in any way at all then it does make sense to leave it in the package but pass in the seed explicitly to the package (by function argument or opts struct field), and the binary (via a flag).

Reply:
I added a flag for it.